### ISSUE TITLE
git completions ignore submodules

### ISSUE THREAD

**Issue by mqudsi** [fish-shell/fish-shell#5786]
It seems the state of submodules is not reported in the same fashion as that of regular files. Our completions are unaware of submodules.

Here's some example output:

```
mqudsi@Blitzkrieg ~/.c/nvim> git reset
Unstaged changes after reset:
M       bundle/repos/github.com/Shougo/dein.vim
mqudsi@Blitzkrieg ~/.c/nvim> git status --porcelain=v2
1 .M SC.. 160000 160000 160000 81779a5aeaa922842172f8d9ca61fa18212719f9 81779a5aeaa922842172f8d9ca61fa18212719f9 bundle/repos/github.com/Shougo/dein.vim
? ftdetect/
? syntax/vssln.vim
mqudsi@Blitzkrieg ~/.c/nvim> git add bundle/repos/github.com/Shougo/dein.vim
mqudsi@Blitzkrieg ~/.c/nvim> git status --porcelain=v2
1 M. S... 160000 160000 160000 81779a5aeaa922842172f8d9ca61fa18212719f9 c802e3aa61d270ea450409bf63c3bdbe90275dde bundle/repos/github.com/Shougo/dein.vim
? ftdetect/
? syntax/vssln.vim
```

**Comment by faho** [comment]
Yeah, this is on purpose. We pass `--ignore-submodules=all` to `git status`, both because that normally only yields the submodule itself, which then isn't usable as an argument, and because it can take ages.

It's possible they now are usable as arguments - I remember testing a bunch of commands, and they just did nothing - so we might need to revisit this.

**Comment by mqudsi** [comment]
What I understand is that the entirety of a submodule is represented as a single entry in the parent repo's index, and it can be treated like a file that can only be staged if its commit has changed. A submodule with changes can't be staged until the changes are committed to the submodule itself; until that point the submodule is in a dirty state but you can't do anything about it from the parent repo.

**Comment by gerph** [comment]
This is quite a frustration for me.

As part of a normal workflow for a project I work on, I would update a submodule, add the files as necessary, commit, maybe push (usual stuff), then cd to the parent directory and try to add the submodule so that the newly updated submodule is present in the index.

If 'MyModule' exists in the directory as the submodule, typing `git add My` will often display 'Module' in grey to indicate that it is a completion if I press cursor-right. If I do press cursor-right, it will complete with that name. That's good, but that's usually only something I do if I know the name would be in my history (as the cursor-right suggestions are predictive based on the environment and your history, so they might not be right).

If I press Tab at that point, I get nothing, which stalls my thoughts and means that I need to think about whether the cursor-right suggestion is right (which it might not be - there might be other names in the suggestion as well as the directory I want to add, so I have to decide to use alt-right to accept one word, or just right to accept the lot and then delete things).

Maybe just changing the logic so that for `git add` a directory is a tab-completion candidate would suffice, as it's not unreasonable to add a directory in a git add (without submodule) and it would avoid the need to include the submodules in the `git status` operation?

**Comment by damnever** [comment]
Is there any chance to add an option to control this?

**Comment by faho** [comment]
Options like that aren't something we add.
